### PR TITLE
Grafana annotation support

### PR DESCRIPTION
**Pull request by yanscalyr**
https://scalyr.myjetbrains.com/youtrack/issue/CT-149
https://grafana.com/docs/grafana/latest/dashboards/annotations/

Support the ability to generate annotations from queries to Scalyr in Grafana using our datasource.

We query Scalyr for any logs matching the user provided query, transform them into the format Grafana expects for annotations, and then pass them along to Grafana which will show them as vertical lines on the user's graphs. Horizontal position determined by timestamp and message given when hovered over given by the `message` field, though both of these are configurable. Optionally the user can configure a field to act as an "end time" for the annotation to show up as an area on the graph.

We have no specific requirements for what features to support, so far we include the ability to chose which fields in the logs will map to the annotation text and timestamps (in milliseconds). Future things to consider supporting is picking a field to grab tags, as well as the ability to chose a unit for the timestamps.

### CODE REVIEW

**Comment by yanscalyr** [comment]
Still need some unit tests but wouldn't mind reviews while I work on that.

File: src/datasource.js [hunk 1]
```diff
@@ -246,6 +277,35 @@ export class GenericDatasource {
...
+   * @param results
+   * @param options
+   * @returns Array
+   */
+  static transformAnnotationResults(results, timeField, timeEndField, textField) {
+    const annotations = [];
+    results.forEach((result) => {
+      const responseObject = {};
+      responseObject.time = Number(result[timeField]) / 1000000;
+      if (!responseObject.time || responseObject.time.length === 0) {
```
Comment (Kami) [inline src/datasource.js:291]: So if ``result[timeField]`` is not defined or similar, ``responseObject.time`` will contain NaN, right?

Also, I wonder why we check for ``time.length``. ``Number`` object doesn't contain ``length`` property, right?

File: src/specs/datasource.test.js [hunk 2]
```diff
@@ -135,4 +148,46 @@ describe('Scalyr datasource tests', () => {
...
+      expect(request.url).toBe('proxied/query');
+      expect(request.method).toBe('POST');
+      const requestBody = JSON.parse(request.data);
+      expect(requestBody.token).toBe('123');
+      expect(requestBody.queryType).toBe('log');
+      expect(requestBody.startTime).toBe(sixHoursAgo.toISOString());
+      expect(requestBody.endTime).toBe(now.toISOString());
+    });
+
+    it('Should transform standard query results to annotations', () => {
```
Comment (Kami) [inline src/specs/datasource.test.js:183]: It would be good to also add test cases for all the edge cases in ``transformAnnotationResults`` function.

File: src/datasource.js (hunk 1 above)
Comment (yanscalyr) [reply src/datasource.js:291]: Yes it can contain `NaN` or `undefined`, this ends up working out with no errors thrown at all and simply not showing the annotation.

You're right about checking length on number making no sense, this was from an earlier iteration.

File: src/partials/annotations.editor.html [hunk 3]
```diff
@@ -0,0 +1,29 @@
...
+
+<div class="gf-form-group">
+	<h6>Field mappings</h6>
+	<div class="gf-form-inline">
+		<div class="gf-form">
+			<span class="gf-form-label">Time</span>
+			<input type="text" class="gf-form-input max-width-14" ng-model='ctrl.annotation.timeField' placeholder="timestamp"></input>
+		</div>
+		<div class="gf-form">
+			<span class="gf-form-label">Time End</span>
```
Comment (oliverhsu77) [inline src/partials/annotations.editor.html:21]: If my understanding is correct, the only time they would have a `timeEnd` field is if they have a special event that contains data for a region annotation - `time` and `timeEnd`.  Unless they have something like this, in most cases we will just use the event `timestamp` field.

**Review by oliverhsu77 — CHANGES_REQUESTED** [review]

File: src/datasource.js [hunk 4]
```diff
@@ -246,6 +277,35 @@ export class GenericDatasource {
...
+   * Transform data returned by time series query into Grafana annotation format.
+   * @param results
+   * @param options
+   * @returns Array
+   */
+  static transformAnnotationResults(results, timeField, timeEndField, textField) {
+    const annotations = [];
+    results.forEach((result) => {
+      const responseObject = {};
+      responseObject.time = Number(result[timeField]) / 1000000;
```
[the remark this answers is not in the document]
Comment (yanscalyr) [reply src/datasource.js:290]: `NaN` gets to the Grafana internals just fine with no errors, but no annotation will show up. I don't see how to better handle that personally so I just left it as is.

File: src/specs/datasource.test.js [hunk 5]
```diff
@@ -135,4 +148,83 @@ describe('Scalyr datasource tests', () => {
...
+      expect(resultEntry.time).toBe(0.012345);
+      expect(resultEntry.timeEnd).toBe(0.054321);
+    });
+
+    it('Should transform standard query results to annotations with bad field names', () => {
+      const transformedResults = GenericDatasource.transformAnnotationResults(results, "missingField", null, null);
+      expect(transformedResults.length).toBe(3);
+      const resultEntry = transformedResults[0];
+      expect(resultEntry.text).toBe(undefined);
+      expect(resultEntry.time).toBe(NaN);
```
[the remark this answers is not in the document]
Comment (yanscalyr) [reply src/specs/datasource.test.js:225]: From my testing Grafana handles bad data quietly, not showing an annotation in this case.

File: src/partials/annotations.editor.html (hunk 3 above)
Comment (yanscalyr) [reply src/partials/annotations.editor.html:21]: That is correct, `timeEnd` defaults to nothing and isn't put into the annotation object at all in that case. Default for `time` is `timestamp`, and we end up with just a vertical line for the annotation.

File: src/datasource.js [hunk 6]
```diff
@@ -59,6 +59,19 @@ export class GenericDatasource {
...
 
+  annotationQuery(options) {
+    const query = this.createLogsQueryForAnnotation(options);
+    return this.backendSrv.datasourceRequest(query)
+      .then( (response) => {
+        const data = response.data;
+        const timeField = options.annotation.timeField || "timestamp"
+        const timeEndField = options.annotation.timeEndField || null
+        const textField = options.annotation.textField || "message"
+        return GenericDatasource.transformAnnotationResults(data.matches, timeField, timeEndField, textField);
```
[the remark this answers is not in the document]
Comment (yanscalyr) [reply src/datasource.js:70]: I hadn't thought of this case, but having just tried it it throws up an error in the Grafana UI with the proper error message received from Scalyr. I believe the proxy layer Grafana has for interacting with datasources has error handling like this. Honestly this seems like the optimal thing to do, so I think we should leave it as is, unless you think otherwise?

[the remark this answers is not in the document]
Comment (Kami) [reply src/datasource.js:70]: Just to add this here for reference / context, since I was also curious - https://grafana.com/docs/grafana/latest/packages_api/runtime/backendsrv/

> By default Grafana will display an error message alert if the remote call fails. If you want to prevent this from happending you need to catch the error thrown by the BackendSrv and set the showErrorAlert = true on the request options object.

So yeah, it looks like that as long as API correctly returns non-200 status code we should be fine (as long as we want to directly show the error returned by the API)

File: src/datasource.js [hunk 7]
```diff
@@ -194,6 +207,24 @@ export class GenericDatasource {
...
+      url: this.url + '/query',
+      method: 'POST',
+      headers: this.headers,
+      data: JSON.stringify({
+        token: this.apiKey,
+        queryType: "log",
+        filter: queryText,
+        startTime: options.range.from.valueOf(),
+        endTime: options.range.to.valueOf(),
+        maxCount: 5000
```
Comment (Kami) [inline src/datasource.js:223]: Is there a situation where we would want this value to be lower / higher (I assume there is some kind of upper limit on the API side)?

File: src/datasource.js (hunk 4 above)
[the remark this answers is not in the document]
Comment (Kami) [reply src/datasource.js:290]: To be on the safe side, one option for us would also be to not add an object to ``annotations`` if both fields are either ``undefined`` / ``NaN`` or if only one field is set (``time``) and it's ``undefined`` / ``NaN``.

Or if only one attribute is ``undefined`` / ``NaN``, not set that attribute on the ``resposneObject``.

File: src/datasource.js (hunk 7 above)
Comment (yanscalyr) [reply src/datasource.js:223]: 5000 is the upper limit of the API side, if we wanted to retrieve more values we could repeat the requests with the continuation token, but as I found out while testing Grafana can't even handle 5000 entries. In fact I should probably lower this, but I don't know what value would be reasonable, possibly 200? I'm unsure.

File: src/partials/annotations.editor.html (hunk 3 above)
Comment (oliverhsu77) [reply src/partials/annotations.editor.html:21]: We can leave this as is, but is sounds like to me for 80-90+% of the cases, the user won't have have a `timeEnd`.  If we don't support `timeEnd`, we could simplify things to where they just need the `message` field and we default to the `timestamp` field for the `time.

File: src/datasource.js (hunk 4 above)
[the remark this answers is not in the document]
Comment (yanscalyr) [reply src/datasource.js:290]: Cases where the `text` field is invalid can still work by just showing the annotation line with no message, but if there is no `time` then yeah there is no reason to send it along.

File: src/datasource.js [hunk 8]
```diff
@@ -246,6 +277,35 @@ export class GenericDatasource {
...
+      if ((!responseObject.text || responseObject.text.length === 0) && result.attributes) {
+        responseObject.text = result.attributes[textField];
+      }
+      if (timeEndField && !(timeEndField.length === 0)) {
+        responseObject.timeEnd = Number(result[timeEndField]) / 1000000;
+        if (!responseObject.timeEnd && result.attributes) {
+          responseObject.timeEnd = Number(result.attributes[timeEndField]) / 1000000;
+        }
+      }
+      annotations.push(responseObject);
```
[the remark this answers is not in the document]
Comment (oliverhsu77) [reply src/datasource.js:309]: Another case to consider:
JavaScript `Number` behavior is:
```
Number('')        // 0
Number(null)      // 0
Number(undefined) // NaN
```
We should also not send if `time === 0`

File: src/datasource.js [hunk 9]
```diff
@@ -246,6 +277,40 @@ export class GenericDatasource {
...
+   * @param results
+   * @param options
+   * @returns Array
+   */
+  static transformAnnotationResults(results, timeField, timeEndField, textField) {
+    const annotations = [];
+    results.forEach((result) => {
+      const responseObject = {};
+      responseObject.time = Number(result[timeField]) / 1000000;
+      if (!responseObject.time && result.attributes) {
```
Comment (oliverhsu77) [inline src/datasource.js:291]: It seems like there may be a logic error:
```
responseObject.time = Number(result[timeField]) / 1000000;
 if (!responseObject.time && result.attributes) {
```
If `timeField` is not in `result`, then `responseObject.time = NaN`.  Then `!responseObject.time` will be `false` b/c `responseObject.time` is not `undefined` but rather `NaN`.

It may make the logic cleaner to do safety checks for `if (result[timeField]) ... else if (result.attributes && result.attributes[timeField]`

**Review by oliverhsu77 — CHANGES_REQUESTED** [review]

File: src/datasource.js (hunk 6 above)
[the remark this answers is not in the document]
Comment (oliverhsu77) [reply src/datasource.js:70]: I did verify the Scalyr api return non-200.

File: src/datasource.js (hunk 9 above)
Comment (yanscalyr) [reply src/datasource.js:291]: `NaN` is also falsey so I think this still works as is, unless I misunderstood what you mean.

Comment (yanscalyr) [reply src/datasource.js:291]: Actually so is `0` so the `time === 0` case is covered here as well.

File: src/datasource.js (hunk 8 above)
[the remark this answers is not in the document]
Comment (Kami) [reply src/datasource.js:309]: One option would be just to check if the value is falsy - this should cover all the scenarios - 0 int, undefined, null, NaN.

But yeah, JavaScript has a bunch of edge cases related to type casting edge cases so in other cases strict comparison is usually preferred.

**Comment by Kami** [comment]
There is a more of a long term wish and not something we can do at this point - but eventually some automated end to end tests would be nice (aka spin up Grafana instance, connect it to some testing Scalyr account and do some basic checks using Selenium or whatever is popular these days).

That would come handy even just from the testing perspective - e.g. docker compose setup which spins up Grafana + related things.

File: src/datasource.js (hunk 8 above)
[the remark this answers is not in the document]
Comment (yanscalyr) [reply src/datasource.js:309]: We currently check for falsyness so is there anything that actually needs changing?

File: src/datasource.js [hunk 10]
```diff
@@ -246,6 +277,40 @@ export class GenericDatasource {
...
+        responseObject.time = Number(result.attributes[timeField]) / 1000000;
+      }
+      if (!responseObject.time) {
+        return; // The `.forEach` equivalent of `continue`
+      }
+
+      responseObject.text = result[textField];
+      if ((!responseObject.text || responseObject.text.length === 0) && result.attributes) {
```
Comment (oliverhsu77) [inline src/datasource.js:299]: Sorry for not bringing this up earlier.  I'm just learning about JavaScript, but I think `(!responseObject.text || responseObject.text.length === 0) && result.attributes)` can be simplified to `(!responseObject.text && result.attributes)` b/c empty string is `falsy`.

File: src/datasource.js [hunk 11]
```diff
@@ -246,6 +277,40 @@ export class GenericDatasource {
...
+   */
+  static transformAnnotationResults(results, timeField, timeEndField, textField) {
+    const annotations = [];
+    results.forEach((result) => {
+      const responseObject = {};
+      responseObject.time = Number(result[timeField]) / 1000000;
+      if (!responseObject.time && result.attributes) {
+        responseObject.time = Number(result.attributes[timeField]) / 1000000;
+      }
+      if (!responseObject.time) {
```
Comment (oliverhsu77) [inline src/datasource.js:294]: It seems like you are trying to avoid some extra processing by returning early if `time` is not defined.  If they have their query correct, `time` should be defined.

From Google, there is no `forEach` continue.  The `return` is returning from the function early, which is slightly hacky.  I think this code is clearer as:
```
if (responseObject.time) {
   annotations.push(responseObject);
}
```
We may do a little more processing getting the `message` field if `time` is not defined, but I think the code is clearer that way.

**Review by oliverhsu77 — CHANGES_REQUESTED** [review]
Sorry, for the multiple rounds.  I'm new to JavaScript.

File: src/datasource.js [hunk 12]
```diff
@@ -246,6 +277,39 @@ export class GenericDatasource {
...
+      if (!responseObject.time && result.attributes) {
+        responseObject.time = Number(result.attributes[timeField]) / 1000000;
+      }
+
+      responseObject.text = result[textField];
+      if (!responseObject.text && result.attributes) {
+        responseObject.text = result.attributes[textField];
+      }
+
+      if (timeEndField && !(timeEndField.length === 0)) {
```
Comment (oliverhsu77) [inline src/datasource.js:300]: Can also remove the `length` check here.

**Review by oliverhsu77 — CHANGES_REQUESTED** [review]

**Review by oliverhsu77 — APPROVED** [review]
LGTM.  Please do a final end-to-end test with all your recent changes if you haven't done so yet.
